nibabel/elemcont.py

```diff
@@ -0,0 +1,188 @@
+        return iter(self._elems)
+
+    def __len__(self):
+        return len(self._elems)
+
+    def __repr__(self):
+        return ('ElemDict(%s)' %
+                ', '.join(['%r=%r' % x for x in self.items()]))
+
+    def update(self, other):
+        if hasattr(other, 'get_elem'):
+            for key in other:
+                self[key] = other.get_elem(key)
+        else:
+            for key, elem in iteritems(other):
+                self[key] = elem
+
+    def get_elem(self, key):
+        return self._elems[key]
+
+
+class ElemList(MutableSequence):
+    '''A list-like container where each value is an "element", which is
+    defined as any object which has a `value` attribute.
+
+    When looking up an item in the list, it is this `value` attribute that
+    is returned. To get the element itself use the `get_elem` method.
+    '''
+    def __init__(self, data=None):
+        self._elems = list()
+        if data is not None:
+            if isinstance(data, self.__class__):
+                for idx in range(len(data)):
+                    self.append(data.get_elem(idx))
+            else:
+                for elem in data:
+                    self.append(elem)
+
+    def _tuple_from_slice(self, slc):
+        '''Get (start, end, step) tuple from slice object.
+        '''
+        (start, end, step) = slc.indices(len(self))
+        # Replace (0, -1, 1) with (0, 0, 1) (misfeature in .indices()).
+        if step == 1:
+          if end < start:
+            end = start
+          step = None
+        if slc.step == None:
+          step = None
+        return (start, end, step)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            return ElemList(self._elems[idx])
+        else:
+            return self._elems[idx].value
+
+    def __setitem__(self, idx, val):
+        if isinstance(idx, slice):
+            (start, end, step) = self._tuple_from_slice(idx)
+            if step != None:
+                # Extended slice
+                indices = range(start, end, step)
+                if len(val) != len(indices):
+                    raise ValueError(('attempt to assign sequence of size %d' +
+                                      ' to extended slice of size %d') %
+                                     (len(value), len(indices)))
+                for j, assign_val in enumerate(val):
+                    self.insert(indices[j], assign_val)
+            else:
+                # Normal slice
+                for j, assign_val in enumerate(val):
+                    self.insert(start + j, assign_val)
+        else:
+            self.insert(idx, val)
+
+    def __delitem__(self, idx):
+        del self._elems[idx]
+
+    def __len__(self):
+        return len(self._elems)
+
+    def __repr__(self):
+        return ('ElemList([%s])' % ', '.join(['%r' % x for x in self]))
+
+    def __add__(self, other):
+        result = self.__class__(self)
+        if isinstance(other, self.__class__):
```

Comment:
I agree `iter_elems` is good for the `ElemDict` class, but this should really be using `enumerate` right?